### PR TITLE
Prevent set_directive from clobbering other keys / safe directive setter

### PR DESCRIPTION
**Pull request by tiran**
This PR combines @frasertweedale PR #1347 with a safe directive setter (https://pagure.io/freeipa/issue/7312)

## Original PR message

`set_directive` only looks for a prefix of the line matching the
given directive (key).  If a directive is encountered for which the
given key is prefix, it will be vanquished.

This occurs in the case of `{ca,kra}.sslserver.cert[req]`; the
`cert` directive gets updated after certificate renewal, and the
`certreq` directive gets clobbered.  This can cause failures later
on during KRA installation, and possibly cloning.

Match the whole directive to avoid this issue.

Fixes: https://pagure.io/freeipa/issue/7288

-----

Cause: corner case.

How to test:

1. ensure `ca.sslserver.certreq=<base64 CSR>` exists in `ca/CS.cfg`.
2. resubmit Certmonger tracking request for `Server-Cert cert-pki-ca` Dogtag system cert.
3. verify that `ca.sslserver.certreq=<base64 CSR>` still exists in `ca/CS.cfg`.

## safe DirectiveSetter

The new context manager ``DirectiveSetter`` avoids several possible issues that can lead to destroyed configuration files.

### CODE REVIEW

File: ipaserver/install/installutils.py [hunk 1]
```diff
@@ -436,6 +438,71 @@ def unquote_directive_value(value, quote_char):
...
+        # rename file and sync directory inode
+        os.rename(name, self.filename)
+        dirfd = os.open(directory, os.O_RDONLY | os.O_DIRECTORY)
+        try:
+            os.fsync(dirfd)
+        finally:
+            os.close(dirfd)
+
+    def set(self, directive, value, quotes=SENTINEL, separator=SENTINEL):
+        if quotes is SENTINEL:
```
Comment (frasertweedale) [inline ipaserver/install/installutils.py]: What's the advantage of a sentinel `object` over plain old `None`?

File: ipaserver/install/installutils.py [hunk 2]
```diff
@@ -480,8 +549,9 @@ def set_directive_lines(quotes, separator, k, v, lines):
         new_line = ''.join([k, separator, v_quoted, '\n'])
 
     found = False
+    matcher = re.compile(r'\s*{}'.format(re.escape(k + separator)))
```
Comment (frasertweedale) [inline ipaserver/install/installutils.py]: :+1:

Obvious, in hindsight :)

File: ipaserver/install/installutils.py [hunk 3]
```diff
@@ -436,6 +438,71 @@ def unquote_directive_value(value, quote_char):
...
+            os.close(dirfd)
+
+    def set(self, directive, value, quotes=SENTINEL, separator=SENTINEL):
+        if quotes is SENTINEL:
+            quotes = self.quotes
+        if separator is SENTINEL:
+            separator = self.separator
+        # materialize lines
+        # set_directive_lines() modify item, shrink or enlage line count
+        self.lines = list(set_directive_lines(
```
Comment (frasertweedale) [inline ipaserver/install/installutils.py]: This exposes another potential optimisation.  We can defer the construction of the `list` and just build
generator upon generator.  Force the `list` just before we write the file, and we will traverse the input
only once.

Not specifically recommending this, just pointing out the possibility :)

File: ipaserver/install/installutils.py [hunk 4]
```diff
@@ -436,6 +438,71 @@ def unquote_directive_value(value, quote_char):
     return unescaped_value
 
 
+SENTINEL = object()
+
+
+class DirectiveSetter(object):
+    """Safe directive setter
+
```
Comment (frasertweedale) [inline ipaserver/install/installutils.py]:
```
PEP-8 errors:
./ipaserver/install/installutils.py:446:1: W293 blank line contains whitespace
```

File: ipaserver/install/installutils.py (hunk 1 above)
Comment (tiran) [reply ipaserver/install/installutils.py]: You can pass in ``None`` to override the default. With ``None`` as default method argument, there is no way to distinguish between ``set(k, v)`` and ``set(k, v, quotes=None)``.

File: ipaserver/install/installutils.py (hunk 3 above)
Comment (tiran) [reply ipaserver/install/installutils.py]: Another optimization idea: Modify the list in-place and use ``None`` as indicator for line removed.

**Comment by tiran** [comment]
I had another idea to make code even more readable:

```
with DirectiveSetter(filename) as ds:
    ds.setlist([
        ('key1', 'value1'),
        ('key2', 'value2'),
        ...
        ('key99', 'value99'),
    ])
```

Also ``SENTINEL`` should rather be ``_SENTINEL``.

File: ipaserver/install/installutils.py (hunk 1 above)
Comment (frasertweedale) [reply ipaserver/install/installutils.py]: But `None` is not a valid value for either `quotes` or `separator` parameters.
I'd rather just use `None` to indicate "existing or default setting" and instead
add check for `isinstance(quotes, bool)` and `isinstance(separator, str) and len(separator) > 0`.

**Comment by frasertweedale** [comment]
@tiran yeah `setlist` is a nice idea.

**Comment by flo-renaud** [comment]
Hi @tiran 
thank you for the patch, works for me.

**Comment by tiran** [comment]
master:

* 2546ef6eb0c6321da810f42d0311b7053be43d62 Prevent set_directive from clobbering other keys
* 1b04718b3c57eed2131db018b9d4e46c8d7f7345 pep8: reduce line lengths in CAInstance.__enable_crl_publish
* c77f3a50d7a8d4b2f1e6fb7c95115a36a4ec6daa installutils: refactor set_directive
* f688b5d8a7ff340a4f358e99bfe219167832359e Add tests for installutils.set_directive
* f4001e1c53a263aa6c7f62385ed394631345a34c Add safe DirectiveSetter context manager
